src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Legacy/HtmlMarkupParser.cs

```diff
@@ -2296,5 +2344,7 @@ private class TagTracker
         public SyntaxList<RazorSyntaxNode> PreviousNodes { get; }
 
         public bool IsWellFormed { get; }
+
+        public ParserContextState? StateBeforeVoidTag { get; init; }
```

Comment:
This is likely not good enough; what if there are multiple? Please add some tests with multiple tags, some of which should be treated as void tags, some of which should not, all within a markup block inside a code block.

Reply:
Good catch, thanks. I'm not yet sure if this will be difficult to fix, need to investigate. But alternatively we could give up this PR but at least emit some better error/warning when trying to use a component named like void tag.

Reply:
Better backtracking support in general will be useful trying to fix some of the nested `}` error recovery scenarios, so I think it would be useful to invest in fixing, even if more complex.

Now, whether it's good _language design_ to allow components to shadow void tags... that's an entirely different question 😆. I would have forbidden it and required an escape syntax of some kind, like C# does for identifiers named after keywords.